### PR TITLE
update `cast` lib to enable map envvar binding, update doc

### PR DESCRIPTION
**Pull request by xvello**
### What does this PR do?

- update the `spf13/cast` version to use the map envvar binding brought by https://github.com/spf13/cast/pull/59
- add `BindEnvVar` for tagger options to extract docker labels & envvars, kubernetes annotations
- update the docker image README, and rework its structure. See it here: https://github.com/DataDog/datadog-agent/tree/xvello/map_in_envvar/Dockerfiles/agent

### Motivation

We had a regression from agent5:

- kubernetes annotations were all collectected by default
- docker label as tags was configurable via an envvar

### CODE REVIEW

File: Dockerfiles/agent/README.md [hunk 1]
```diff
@@ -4,25 +4,65 @@ This is how the official agent 6 image available [here](https://hub.docker.com/r
...
+
+#### Tagging
+
+We automatically collect common tags from [Docker](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/docker_extract.go), [Kubernetes](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/kubelet_extract.go), [ECS](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/ecs_extract.go), [Swarm, Mesos, Nomad and Rancher](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/docker_extract.go), and allow you to extract even more tags with the following options:
+
+- `DD_DOCKER_LABELS_AS_TAGS` : extract docker container labels
+- `DD_DOCKER_ENV_AS_TAGS` : extract docker container environment variables
+- `DD_KUBERNETES_POD_LABELS_AS_TAGS` : extract pod labels
+
+You can either define them in your custom `datadog.yaml`, or set them as JSON maps in these envvars. The map key is the source (label/envvar) name, and the map value the datadog tag name. If you prefix the tag name with +, it will only be added to high cardinality metrics (docker check). For example:
```
Comment (hkaj) [inline Dockerfiles/agent/README.md]: given we're moving to handling high-card tags in the backend, I don't think we need to document the last part

**Review by hkaj — CHANGES_REQUESTED** [review]
nice, left 1 note, lgtm otherwise.

File: Dockerfiles/agent/README.md [hunk 2]
```diff
@@ -4,25 +4,65 @@ This is how the official agent 6 image available [here](https://hub.docker.com/r
...
+We automatically collect common tags from [Docker](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/docker_extract.go), [Kubernetes](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/kubelet_extract.go), [ECS](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/ecs_extract.go), [Swarm, Mesos, Nomad and Rancher](https://github.com/DataDog/datadog-agent/blob/master/pkg/tagger/collectors/docker_extract.go), and allow you to extract even more tags with the following options:
+
+- `DD_DOCKER_LABELS_AS_TAGS` : extract docker container labels
+- `DD_DOCKER_ENV_AS_TAGS` : extract docker container environment variables
+- `DD_KUBERNETES_POD_LABELS_AS_TAGS` : extract pod labels
+
+You can either define them in your custom `datadog.yaml`, or set them as JSON maps in these envvars. The map key is the source (label/envvar) name, and the map value the datadog tag name.
+
+```
+DD_KUBERNETES_POD_LABELS_AS_TAGS = {\"app\": \"kube_app\", \"release\": \"helm_release\"}"
```
Comment (JulienBalestra) [inline Dockerfiles/agent/README.md]: I think we shouldn't add spaces between the `=`.
I could be like:
```bash
DD_KUBERNETES_POD_LABELS_AS_TAGS='{"app": "kube_app", "release": "helm_release"}'
echo $DD_KUBERNETES_POD_LABELS_AS_TAGS 
{"app": "kube_app", "release": "helm_release"}
```

File: Dockerfiles/agent/README.md [hunk 3]
```diff
@@ -4,25 +4,65 @@ This is how the official agent 6 image available [here](https://hub.docker.com/r
...
+DD_DOCKER_LABELS_AS_TAGS = "{\"com.docker.compose.service\": \"service_name\", \"com.docker.compose.project\": \"project_name\"}"
```
Comment (JulienBalestra) [inline Dockerfiles/agent/README.md]: ditto here, and we can replace the edge " by ' as we don't have some variable inside this string.

**Review by JulienBalestra — COMMENTED** [review]
I have some open questions on the envvar setting.

File: Dockerfiles/agent/README.md (hunk 3 above)
Comment (JulienBalestra) [reply Dockerfiles/agent/README.md]: Then get rid of the escape `\"`

Comment (xvello) [reply Dockerfiles/agent/README.md]: Would quoting the envvar value with single quotes work in a kubernetes daemonset spec?
As I already had a few tickets about this for AD, I went with the most copy-paste proof syntax

Comment (JulienBalestra) [reply Dockerfiles/agent/README.md]: Yes we can do something like:
```yaml
 - name: COOL_ENV
   value: '{"ok": "val"}'
```

And as expect observe in the container:
```bash
env 
```
```text
...
COOL_ENV={"ok": "val"}
...
```

Comment (JulienBalestra) [reply Dockerfiles/agent/README.md]: The spaces in the env as json string are mandatory for the env binding ?

**Review by JulienBalestra — DISMISSED** [review]
:+1:

**Pull request merged**